python/smqtk/web/nn_descriptor_service/__init__.py

```diff
@@ -0,0 +1,231 @@
+                message = "URI resolution issue: %s" % str(ex)
+
+            if de:
+                try:
+                    descriptor = self.descriptor_generator_inst.\
+                        compute_descriptor(de, self.descr_elem_factory)
+
+                    # TODO: Clarify if converting to 1D is necessary
+                    # (it seems for Caffe descriptor)
+                    descriptor.set_vector(descriptor.vector().flatten())
+                except RuntimeError, ex:
+                    message = "Descriptor extraction failure: %s" % str(ex)
+                except ValueError, ex:
+                    message = "Data content type issue: %s" % str(ex)
+
+            # fail here if de is None
+            # Default is 8
+            num_neighbors = flask.request.args.get("num_neighbors", 8)
+            neighbors, _ = self.nn_index.nn(descriptor, n=num_neighbors)
+
+            # TODO: Return the optional descriptor vector for the neighbors
+            return flask.jsonify({
+                "success": descriptor is not None,
+                "message": message,
+                "neighbors": [n.uuid() for n in neighbors],
+                "reference_uri": uri
+            })
+
+    def get_config(self):
+        return self.json_config
+
+    def resolve_data_element(self, uri):
+        """
+        Given the URI to some data, resolve it down to a DataElement instance.
+
+        :raises ValueError: Issue with the given URI regarding either URI source
+            resolution or data resolution.
+
+        :param uri: URI to data
+        :type uri: str
+        :return: DataElement instance wrapping given URI to data.
+        :rtype: smqtk.representation.DataElement
+
+        """
+        # Resolve URI into appropriate DataElement instance
+        if uri[:7] == "file://":
+            self.log.debug("Given local disk filepath")
+            filepath = uri[7:]
+            if not os.path.isfile(filepath):
+                raise ValueError("File URI did not point to an existing file "
+                                 "on disk.")
+            else:
+                de = DataFileElement(filepath)
+
+        elif uri[:9] == "base64://":
+            self.log.debug("Given base64 string")
+            content_type = flask.request.args.get('content_type', None)
+            self.log.debug("Content type: %s", content_type)
+            if not content_type:
+                raise ValueError("No content-type with given base64 data")
+            else:
+                b64str = uri[9:]
+                de = DataMemoryElement.from_base64(b64str, content_type)
+
+        else:
+            self.log.debug("Given URL")
+            try:
+                de = DataUrlElement(uri)
+            except requests.HTTPError, ex:
+                raise ValueError("Failed to initialize URL element due to "
+                                 "HTTPError: %s" % str(ex))
+
+        return de
+
+APPLICATION_CLASS = NearestNeighborServiceServer
```

Comment:
Technically shouldn't need this any more with recent (ish) plugin discovery updates, but its fine the way it is (check out the get_plugins doc-strings).